clvm-traits/src/from_clvm.rs

```diff
@@ -60,113 +83,116 @@ clvm_primitive!(i128);
 clvm_primitive!(usize);
 clvm_primitive!(isize);
 
-impl<A, B> FromClvm for (A, B)
+impl<N, A, B> FromClvm<N> for (A, B)
 where
-    A: FromClvm,
-    B: FromClvm,
+    A: FromClvm<N>,
+    B: FromClvm<N>,
 {
-    fn from_clvm(a: &Allocator, ptr: NodePtr) -> Result<Self> {
-        match a.sexp(ptr) {
-            SExp::Pair(first, rest) => Ok((A::from_clvm(a, first)?, B::from_clvm(a, rest)?)),
-            SExp::Atom => Err(Error::ExpectedCons(ptr)),
-        }
+    fn from_clvm(decoder: &impl ClvmDecoder<Node = N>, node: N) -> Result<Self, FromClvmError> {
+        let (first, rest) = decoder.decode_pair(&node)?;
+        let first = A::from_clvm(decoder, first)?;
+        let rest = B::from_clvm(decoder, rest)?;
+        Ok((first, rest))
     }
 }
 
-impl FromClvm for () {
-    fn from_clvm(a: &Allocator, ptr: NodePtr) -> Result<Self> {
-        if nullp(a, ptr) {
+impl<N> FromClvm<N> for () {
+    fn from_clvm(decoder: &impl ClvmDecoder<Node = N>, node: N) -> Result<Self, FromClvmError> {
+        let bytes = decoder.decode_atom(&node)?;
+        if bytes.is_empty() {
             Ok(())
         } else {
-            Err(Error::ExpectedNil(ptr))
+            Err(FromClvmError::WrongAtomLength {
+                expected: 0,
+                found: bytes.len(),
+            })
         }
     }
 }
 
-impl<T, const N: usize> FromClvm for [T; N]
+impl<N, T, const LEN: usize> FromClvm<N> for [T; LEN]
 where
-    T: FromClvm,
+    T: FromClvm<N>,
 {
-    fn from_clvm(a: &Allocator, mut ptr: NodePtr) -> Result<Self> {
-        let mut items = Vec::with_capacity(N);
+    fn from_clvm(decoder: &impl ClvmDecoder<Node = N>, mut node: N) -> Result<Self, FromClvmError> {
+        let mut items = Vec::with_capacity(LEN);
```

Comment:
Not important.
I suspect that this would be more efficient and maybe more straight-forward code if this was allocated as an array on the stack, then filled in one item at a time with a `for`-loop.

I suppose leaving items uninitialized might add some complexity.

Reply:
Would this require unsafe code? As long as we're fine with introducing some unsafe into the repo, this does seem like a relevant optimization

Reply:
yeah, it would probably be something like this: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element